File: src/consent/consent.service.ts
import {
  Injectable,
  NotFoundException,
  BadRequestException,
  ForbiddenException,
} from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model, Types } from 'mongoose';
import { Consent, ConsentDocument } from './schemas/consent.schema';
import { CreateConsentDto } from './dto/create-consent.dto';
import { Student, StudentDocument } from '../students/schemas/student.schema';
import { UserRole } from '../users/schemas/user.schema';

@Injectable()
export class ConsentService {
  constructor(
    @InjectModel(Consent.name) private consentModel: Model<ConsentDocument>,
    @InjectModel(Student.name) private studentModel: Model<StudentDocument>,
  ) {}

  /**
   * Crear o actualizar el consentimiento general de un estudiante
   */
  async createOrUpdateConsent(
    authenticatedUserId: string,
    createConsentDto: CreateConsentDto,
    ipAddress?: string,
    userAgent?: string,
  ): Promise<ConsentDocument> {
    const { allowsDataSharing, comments } = createConsentDto;

    if (!Types.ObjectId.isValid(authenticatedUserId)) {
      throw new BadRequestException('ID de usuario inválido.');
    }

    // 1. Buscar el estudiante asociado al usuario autenticado
    const student = await this.studentModel
      .findOne({ userId: new Types.ObjectId(authenticatedUserId) })
      .populate('carreraId', 'name')
      .exec();

    if (!student) {
      throw new NotFoundException(
        'No se encontró un estudiante asociado a este usuario.',
      );
    }

    // 2. Preparar datos del consentimiento
    const consentData = {
      studentId: student._id,
      allowsDataSharing,
      consentDate: new Date(),
      studentRut: student.rut,
      studentName: `${student.nombres} ${student.apellidos}`,
      studentCareer: (student.carreraId as any)?.name || 'Carrera no especificada',
      comments,
      registeredBy: new Types.ObjectId(authenticatedUserId),
      ipAddress,
      userAgent,
      isActive: true,
    };

    // 3. Upsert: actualizar existente o crear nuevo
    const consent = await this.consentModel
      .findOneAndUpdate(
        { studentId: student._id },
        { $set: consentData },
        { new: true, upsert: true, runValidators: true },
      )
      .exec();

    return consent;
  }

  /**
   * Obtener el consentimiento activo de un estudiante por su ID de usuario
   */
  async getConsentByUserId(userId: string): Promise<ConsentDocument | null> {
    if (!Types.ObjectId.isValid(userId)) {
      throw new BadRequestException('ID de usuario inválido.');
    }

    const student = await this.studentModel
      .findOne({ userId: new Types.ObjectId(userId) })
      .exec();

    if (!student) {
      return null;
    }

    return this.consentModel
      .findOne({ studentId: student._id, isActive: true })
      .exec();
  }

  /**
   * Obtener el consentimiento activo de un estudiante por su Student ID
   */
  async getConsentByStudentId(studentId: string): Promise<ConsentDocument | null> {
    if (!Types.ObjectId.isValid(studentId)) {
      throw new BadRequestException('ID de estudiante inválido.');
    }

    return this.consentModel
      .findOne({ studentId: new Types.ObjectId(studentId), isActive: true })
      .exec();
  }

  /**
   * Verificar si un usuario puede ver información sensible de un estudiante
   */
  async canViewSensitiveData(
    studentId: string,
    viewerRole: UserRole,
    viewerUserId?: string
  ): Promise<boolean> {
    // Coordinadora y Educadora Social: SIEMPRE pueden ver información sensible
    if (viewerRole === UserRole.COORDINADOR || viewerRole === UserRole.EDUCADORA_SOCIAL) {
      return true;
    }

    // El propio estudiante siempre puede ver su información
    if (viewerRole === UserRole.ESTUDIANTE && viewerUserId) {
      const student = await this.studentModel
        .findOne({ userId: new Types.ObjectId(viewerUserId) })
        .exec();
      if (student && student._id.toString() === studentId) {
        return true;
      }
    }

    // Para otros roles (docentes, jefes, etc.): verificar consentimiento
    const consent = await this.getConsentByStudentId(studentId);
    return consent?.allowsDataSharing || false;
  }

  /**
   * Verificar si un usuario puede ver documentos de un estudiante
   */
  async canViewDocuments(
    studentId: string,
    viewerRole: UserRole,
    viewerUserId?: string
  ): Promise<boolean> {
    // Solo Coordinadora y Educadora Social pueden ver documentos
    // Y solo si el estudiante tiene consentimiento activo
    if (viewerRole === UserRole.COORDINADOR || viewerRole === UserRole.EDUCADORA_SOCIAL) {
      const consent = await this.getConsentByStudentId(studentId);
      return consent?.allowsDataSharing || false;
    }

    // El propio estudiante siempre puede ver sus documentos
    if (viewerRole === UserRole.ESTUDIANTE && viewerUserId) {
      const student = await this.studentModel
        .findOne({ userId: new Types.ObjectId(viewerUserId) })
        .exec();
      return student && student._id.toString() === studentId;
    }

    return false;
  }

  /**
   * Verificar si un usuario puede ver ajustes académicos de un estudiante
   */
  async canViewAdjustments(
    studentId: string,
    viewerRole: UserRole,
    viewerUserId?: string
  ): Promise<boolean> {
    // Coordinadora y Educadora Social: SIEMPRE pueden ver ajustes
    if (viewerRole === UserRole.COORDINADOR || viewerRole === UserRole.EDUCADORA_SOCIAL) {
      return true;
    }

    // Docentes: SIEMPRE pueden ver ajustes (sin necesidad de consentimiento)
    // Pero solo pueden ver diagnóstico si hay consentimiento
    if (viewerRole === UserRole.DOCENTE) {
      return true;
    }

    // Jefes de carrera y departamento: pueden ver ajustes
    if (viewerRole === UserRole.JEFE_CARRERA || viewerRole === UserRole.JEFE_DEPARTAMENTO) {
      return true;
    }

    // El propio estudiante siempre puede ver sus ajustes
    if (viewerRole === UserRole.ESTUDIANTE && viewerUserId) {
      const student = await this.studentModel
        .findOne({ userId: new Types.ObjectId(viewerUserId) })
        .exec();
      return student && student._id.toString() === studentId;
    }

    return false;
  }

  /**
   * Revocar el consentimiento de un estudiante
   */
  async revokeConsent(
    authenticatedUserId: string,
    reason?: string,
  ): Promise<ConsentDocument> {
    const consent = await this.getConsentByUserId(authenticatedUserId);
    
    if (!consent) {
      throw new NotFoundException('No se encontró un consentimiento activo.');
    }

    consent.allowsDataSharing = false;
    consent.revokedAt = new Date();
    consent.revocationReason = reason;
    
    return consent.save();
  }

  /**
   * Obtener estadísticas de consentimientos (para reportes administrativos)
   */
  async getConsentStats(): Promise<{
    total: number;
    withConsent: number;
    withoutConsent: number;
    percentageWithConsent: number;
  }> {
    const total = await this.consentModel.countDocuments({ isActive: true });
    const withConsent = await this.consentModel.countDocuments({ 
      isActive: true, 
      allowsDataSharing: true 
    });
    const withoutConsent = total - withConsent;
    const percentageWithConsent = total > 0 ? (withConsent / total) * 100 : 0;

    return {
      total,
      withConsent,
      withoutConsent,
      percentageWithConsent: Math.round(percentageWithConsent * 100) / 100,
    };
  }

  /**
   * Listar todos los consentimientos (para administradores)
   */
  async findAll(): Promise<ConsentDocument[]> {
    return this.consentModel
      .find({ isActive: true })
      .sort({ consentDate: -1 })
      .exec();
  }
}

File: incluye_app/lib/services/consent_service.dart
// services/consent_service.dart
import 'api_service.dart';

class ConsentService {
  // Crear o actualizar consentimiento del estudiante autenticado
  static Future<Map<String, dynamic>?> createOrUpdateConsent({
    required bool allowsDataSharing,
    String? comments,
  }) async {
    try {
      final token = await ApiService.getToken();
      if (token == null) throw Exception('Token no disponible');

      final response = await ApiService.dio.post(
        '/consents',
        data: {
          'allowsDataSharing': allowsDataSharing,
          if (comments != null && comments.isNotEmpty) 'comments': comments,
        },
      );
      
      if (response.statusCode == 201 || response.statusCode == 200) {
        return response.data;
      }
      return null;
    } catch (e) {
      ApiService.handleApiError('Crear/actualizar consentimiento', e);
      return null;
    }
  }

  // Obtener consentimiento del estudiante autenticado
  static Future<Map<String, dynamic>?> getMyConsent() async {
    try {
      final token = await ApiService.getToken();
      if (token == null) throw Exception('Token no disponible');

      final response = await ApiService.dio.get('/consents/my-consent');
      
      if (response.statusCode == 200) {
        return response.data;
      }
      return null;
    } catch (e) {
      ApiService.handleApiError('Obtener mi consentimiento', e);
      return null;
    }
  }

  // Revocar consentimiento del estudiante autenticado
  static Future<bool> revokeConsent({String? reason}) async {
    try {
      final token = await ApiService.getToken();
      if (token == null) throw Exception('Token no disponible');

      final response = await ApiService.dio.patch(
        '/consents/revoke',
        data: {
          if (reason != null && reason.isNotEmpty) 'reason': reason,
        },
      );
      
      return response.statusCode == 200;
    } catch (e) {
      ApiService.handleApiError('Revocar consentimiento', e);
      return false;
    }
  }

  // Obtener consentimiento de un estudiante específico (solo roles autorizados)
  static Future<Map<String, dynamic>?> getStudentConsent(String studentId) async {
    try {
      final token = await ApiService.getToken();
      if (token == null) throw Exception('Token no disponible');

      final response = await ApiService.dio.get('/consents/student/$studentId');
      
      if (response.statusCode == 200) {
        return response.data;
      }
      return null;
    } catch (e) {
      ApiService.handleApiError('Obtener consentimiento del estudiante', e);
      return null;
    }
  }

  // Obtener todos los consentimientos (solo administradores)
  static Future<List<dynamic>?> getAllConsents() async {
    try {
      final token = await ApiService.getToken();
      if (token == null) throw Exception('Token no disponible');

      final response = await ApiService.dio.get('/consents/all');
      
      if (response.statusCode == 200) {
        return response.data is List ? response.data : [response.data];
      }
      return null;
    } catch (e) {
      ApiService.handleApiError('Obtener todos los consentimientos', e);
      return null;
    }
  }

  // Obtener estadísticas de consentimientos (solo administradores)
  static Future<Map<String, dynamic>?> getConsentStats() async {
    try {
      final token = await ApiService.getToken();
      if (token == null) throw Exception('Token no disponible');

      final response = await ApiService.dio.get('/consents/stats');
      
      if (response.statusCode == 200) {
        return response.data;
      }
      return null;
    } catch (e) {
      ApiService.handleApiError('Obtener estadísticas de consentimientos', e);
      return null;
    }
  }

  // Verificar si un estudiante tiene consentimiento activo
  static Future<bool> hasActiveConsent(String studentId) async {
    final consent = await getStudentConsent(studentId);
    return consent != null && 
           consent['isActive'] == true && 
           consent['allowsDataSharing'] == true;
  }

  // Verificar permisos basados en consentimiento y rol
  static bool canAccessData({
    required Map<String, dynamic>? consent,
    required String userRole,
    required String dataType, // 'documents', 'diagnosis', 'adjustments'
  }) {
    // El estudiante siempre puede ver sus propios datos
    if (userRole == 'estudiante') return true;

    // Sin consentimiento, solo se permiten ajustes
    if (consent == null || consent['allowsDataSharing'] != true) {
      return dataType == 'adjustments';
    }

    // Con consentimiento
    switch (userRole) {
      case 'coordinadora':
      case 'educadora':
        return true; // Acceso completo
      case 'docente':
        return dataType != 'documents'; // No documentos
      default:
        return dataType == 'adjustments'; // Solo ajustes
    }
  }
}
